### PR TITLE
(#1240) Memorize Selected Scenario in Scenario Selection Window

### LINKED ISSUE #1240: Scenario selection window, memorized selection
In the original game, when the user selects a scenario from a list according to the size of the scenarios and then the user selects the scenarios of all sizes with the "ALL" selection button, the scenario remains selected in the list displayed (in the example the user selects the small scenarios using the "S" selection button and then selects the "0 Test 06" scenario from the list).

In fheroes2, the scenario does not stay selected.

In the original game:

Scenario selection 07

Scenario selection 08

In fheroes2:

Scenario selection 09

Scenario selection 10

### PR DESCRIPTION
**Pull request by HaveFun91**
Fix #1240

### CODE REVIEW

**Comment by oleg-derevenetz** [comment]
Hi @HaveFun91 hint: to indicate that this PR should close one or more issues upon its merge, refer to these issues in the PR description using the following syntax:

fix #ISSUE_1_ID
fix #ISSUE_2_ID

or

close #ISSUE_1_ID
close #ISSUE_2_ID

**Comment by HaveFun91** [comment]
@ihhub or @oleg-derevenetz feel free to review - already rebased on fresh master :)

File: src/fheroes2/dialog/dialog_selectscenario.cpp [hunk 1]
```diff
@@ -190,6 +189,12 @@ namespace
         const int32_t centerTransform = areaWidth > textWidth ? ( areaWidth - textWidth ) / 2 : 0;
         return startCoordX + centerTransform;
     }
+
+    void SwitchPressedMapSizeButtons( fheroes2::ButtonBase *& currentPressedButton, fheroes2::Button & newlyPressedButton )
+    {
+        currentPressedButton = &newlyPressedButton;
+        currentPressedButton->press();
+    }
```
Comment (ihhub) [inline src/fheroes2/dialog/dialog_selectscenario.cpp:197]: Will this construction work?
```cpp
    fheroes2::ButtonBase * SwitchPressedMapSizeButtons( fheroes2::Button & newlyPressedButton )
    {
        newlyPressedButton.press();
        return &newlyPressedButton;
    }
...

currentPressedButton = SwitchPressedMapSizeButtons( newlyPressedButton );
```

My concern is `*&` usage which is very tricky.

File: src/fheroes2/dialog/dialog_selectscenario.cpp [hunk 2]
```diff
@@ -258,6 +263,30 @@ void ScenarioListBox::_renderMapName( const Maps::FileInfo & info, bool selected
     mapName.draw( xCoordinate, yCoordinate, display );
 }
 
+void ScenarioListBox::SelectMapSize( MapsFileInfoList & mapsList, const Maps::mapsize_t selectedSize_ )
+{
+    const fheroes2::Sprite & originalSlider = fheroes2::AGG::GetICN( ICN::ESCROLL, 3 );
+    const fheroes2::Image updatedScrollbarSlider = fheroes2::generateScrollbarSlider( originalSlider, false, 140, 9, static_cast<int32_t>( mapsList.size() ),
+                                                                                      { 0, 0, originalSlider.width(), 8 }, { 0, 7, originalSlider.width(), 8 } );
+    this->setScrollBarImage( updatedScrollbarSlider );
```
Comment (ihhub) [inline src/fheroes2/dialog/dialog_selectscenario.cpp:271]: Usage of `this->` is redundant. Please remove it here and in other places.

File: src/fheroes2/dialog/dialog_selectscenario.cpp [hunk 3]
```diff
@@ -668,4 +607,4 @@ const Maps::FileInfo * Dialog::SelectScenario( const MapsFileInfoList & all )
     }
 
     return nullptr;
-}
+}
```
Comment (ihhub) [inline src/fheroes2/dialog/dialog_selectscenario.cpp:610]: Please add an empty line at the end of this file.

File: src/fheroes2/dialog/dialog_selectscenario.h [hunk 4]
```diff
@@ -68,6 +69,7 @@ class ScenarioListBox : public Interface::ListBox<Maps::FileInfo>
     bool selectOk;
 
 private:
+    Maps::mapsize_t selectedSize{ Maps::ZERO };
```
Comment (ihhub) [inline src/fheroes2/dialog/dialog_selectscenario.h:72]: I suggest to do this way:
- set an explicit data type for the enumeration:
```cpp
enum mapsize_t : int
```
- use **int** here and in `SelectMapSize()` method  and remove `#include "maps.h"`

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @HaveFun91 , I left few suggestions here. Could you please take a look at them?

File: src/fheroes2/dialog/dialog_selectscenario.h (hunk 4 above)
Comment (HaveFun91) [reply src/fheroes2/dialog/dialog_selectscenario.h:72]: I'm afraid I cannot remove maps.h as this is also used in `dialog_selectscenario.cpp` and i'm getting compilation errors. Other than this - changes applied.

Comment (ihhub) [reply src/fheroes2/dialog/dialog_selectscenario.h:72]: Sure.

File: src/fheroes2/dialog/dialog_selectscenario.h [hunk 5]
```diff
@@ -68,6 +69,8 @@ class ScenarioListBox : public Interface::ListBox<Maps::FileInfo>
     bool selectOk;
 
 private:
+    enum mapsize_t : int;
```
Comment (ihhub) [inline src/fheroes2/dialog/dialog_selectscenario.h:72]: I actually meant to change the enumeration in maps.h file, not to add a line here :)

**Comment by ihhub** [comment]
Hi @LeHerosInconnu, can you please verify this pull request?

**Comment by LeHerosInconnu** [comment]
Hello @ihhub,

> Hi @LeHerosInconnu, can you please verify this pull request?

Works great, and I would say better than in the original game. :)

**Comment by ihhub** [comment]
@HaveFun91 , huge thanks for fixing this issue!

**Pull request merged**